api/service_principals.go

```diff
@@ -14,7 +20,67 @@ type ServicePrincipalClient interface {
 	DeleteServicePrincipal(ctx context.Context, spObjectID string, permanentlyDelete bool) error
 }
 
-type ServicePrincipal struct {
-	ObjectID string
-	AppID    string
+func (c *AppClient) CreateServicePrincipal(ctx context.Context, appID string, startDate time.Time, endDate time.Time) (string, string, error) {
+	spReq := models.NewServicePrincipal()
+	spReq.SetAppId(&appID)
+
+	sp, err := c.client.ServicePrincipals().Post(ctx, spReq, nil)
+	if err != nil {
+		return "", "", err
+	}
+
+	spID := sp.GetId()
+
+	passwordReq := serviceprincipals.NewItemAddPasswordPostRequestBody()
+	passwordCredential := models.NewPasswordCredential()
+	passwordCredential.SetStartDateTime(&startDate)
+	passwordCredential.SetEndDateTime(&endDate)
+
+	passwordReq.SetPasswordCredential(passwordCredential)
+
+	password, err := c.client.ServicePrincipals().ByServicePrincipalId(*spID).AddPassword().Post(context.Background(), passwordReq, nil)
+
+	if err != nil {
+		e := c.DeleteServicePrincipal(ctx, *spID, false)
+		merr := multierror.Append(err, e)
+		return "", "", merr.ErrorOrNil()
+	}
+	return *spID, *password.GetSecretText(), nil
+}
+
+func (c *AppClient) DeleteServicePrincipal(ctx context.Context, spObjectID string, permanentlyDelete bool) error {
+	err := c.client.ServicePrincipals().ByServicePrincipalId(spObjectID).Delete(ctx, nil)
+
+	if permanentlyDelete {
+		e := c.client.Directory().DeletedItems().ByDirectoryObjectId(spObjectID).Delete(ctx, nil)
+		merr := multierror.Append(err, e)
+		return merr.ErrorOrNil()
+	}
+
+	return err
+}
+
+func (c *AppClient) ListServicePrincipals(ctx context.Context, spObjectID string) ([]models.ServicePrincipalable, error) {
+	filter := fmt.Sprintf("appId eq '%s'", spObjectID)
+	requestParameters := &serviceprincipals.ServicePrincipalsRequestBuilderGetQueryParameters{
+		Filter: &filter,
+	}
+
+	headers := abstractions.NewRequestHeaders()
+	headers.Add("ConsistencyLevel", "eventual")
+
+	configuration := &serviceprincipals.ServicePrincipalsRequestBuilderGetRequestConfiguration{
+		Headers:         headers,
+		QueryParameters: requestParameters,
+	}
+
+	spList, err := c.client.ServicePrincipals().Get(ctx, configuration)
+	if err != nil {
+		return nil, err
+	}
+	return spList.GetValue(), nil
+}
+
+func (c *AppClient) GetServicePrincipalByID(ctx context.Context, spObjectID string) (models.ServicePrincipalable, error) {
+	return c.client.ServicePrincipals().ByServicePrincipalId(spObjectID).Get(ctx, nil)
```

Comment:
It looks like these are only used in test code. I'd suggest to keep them in the test code to keep the interface minimal and have a clean separation between the test code and the plugin code.

Reply:
I was having trouble keeping the interface for service principals to the minimal (currently at two methods) when I attempted to move these methods to the test file and define a test interface. I think the moving of these to a test file would require 1 additional method at the least to the interface — something like `GetSPClient` that can be called from the test file to convert a service principal client to a `testSPClient` interface that supports the additional two methods.

Maybe I was looking at things in the wrong way, wrangling all of the different interfaces might have led me down a different path. Would be open to your thoughts on what you think the test file interfaces could look like while maintaining the minimal interface! 🙏🏼  Since these 2 functions are only being used in acceptance tests, I mainly need access to the `MSGraphClient.client *msgraphsdgo.GraphServiceClient`, which seems hard to do without adding an interface method.

I did clean up a little bit of the implementation by adding internal types and not relying on `models.ServicePrincipalable`

Reply:
Thanks for the explanation, Vinay! I can see how the package structure and various client interfaces would make this difficult. I think a method that allows us to get the underlying client would work. I also think it's okay to not do this now.